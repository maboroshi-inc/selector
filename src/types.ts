/**
 * 適合要素型
 */
export type MatchedElement<
  T extends Selector,
  E extends Element
> = T extends keyof HTMLElementTagNameMap
  ? HTMLElementTagNameMap[T]
  : T extends keyof SVGElementTagNameMap
  ? SVGElementTagNameMap[T]
  : E

/**
 * セレクター型
 */
export type Selector =
  | keyof HTMLElementTagNameMap
  | keyof SVGElementTagNameMap
  | string
