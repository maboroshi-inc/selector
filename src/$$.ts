/**
 * セレクターにマッチする全ての要素を返す
 *
 * @param selector セレクター文字列
 * @param ctx コンテキスト要素
 * @returns `ctx` から `selector` にマッチする全ての要素を返す
 */
export function $$<E extends Element = Element>(
  selector: string,
  ctx: ParentNode = document
): NodeListOf<E> {
  return ctx.querySelectorAll<E>(selector)
}
