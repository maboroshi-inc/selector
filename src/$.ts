/**
 * セレクターにマッチする最初の要素を返す
 *
 * @param selector セレクター文字列
 * @param ctx コンテキスト要素
 * @returns `ctx` から `selector` にマッチする最初の要素を返す
 */
export function $<E extends Element = Element>(
  selector: string,
  ctx: Document | Element = document
): E | null {
  return ctx.querySelector(selector)
}
