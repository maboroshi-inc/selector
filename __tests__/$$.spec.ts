import { $, $$ } from '../src'

describe('$$()', () => {
  it('`document`から全ての`p`を返す', () => {
    document.body.innerHTML = `<div><p>hoge</p><p>fuga</p><p>piyo</p></div>`

    expect(Array.from($$('p'))).toMatchSnapshot()
  })

  it('`#context`から全ての`span`を返す', () => {
    document.body.innerHTML = `<div><p><span>hoge</span><span>hoge</span><span>hoge</span></p><p id="context"><span>fuga</span><span>fuga</span><span>fuga</span></p><p><span>piyo</span><span>piyo</span><span>piyo</span></p></div>`

    const contextEl = $<HTMLParagraphElement>('#context')
    expect(contextEl && Array.from($$('span', contextEl))).toMatchSnapshot()
  })
})
