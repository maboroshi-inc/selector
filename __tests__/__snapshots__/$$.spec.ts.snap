// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`$$() \`#context\`から全ての\`span\`を返す 1`] = `
Array [
  <span>
    fuga
  </span>,
  <span>
    fuga
  </span>,
  <span>
    fuga
  </span>,
]
`;

exports[`$$() \`document\`から全ての\`p\`を返す 1`] = `
Array [
  <p>
    hoge
  </p>,
  <p>
    fuga
  </p>,
  <p>
    piyo
  </p>,
]
`;
